src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Extensions.cs

```diff
@@ -0,0 +1,539 @@
+
+            if (!sawUnderlyingType)
+            {
+                diagnostics.Add(ErrorCode.ERR_ExtensionMissingUnderlyingType, Locations.FirstOrNone(), this);
+            }
+            else if (underlyingType is not null)
+            {
+                Debug.Assert(underlyingTypeLocation != null);
+
+                if (!this.IsNoMoreVisibleThan(underlyingType, ref useSiteInfo))
+                {
+                    diagnostics.Add(ErrorCode.ERR_BadVisUnderlyingType, underlyingTypeLocation, this, underlyingType);
+                }
+
+                if (underlyingType.HasFileLocalTypes() && !this.HasFileLocalTypes())
+                {
+                    diagnostics.Add(ErrorCode.ERR_FileTypeUnderlying, underlyingTypeLocation, underlyingType, this);
+                }
+            }
+
+            var baseExtensions = baseExtensionsBuilder.ToImmutableAndFree();
+            foreach (var baseExtension in baseExtensions)
+            {
+                if (!baseExtension.IsAtLeastAsVisibleAs(this, ref useSiteInfo))
+                {
+                    diagnostics.Add(ErrorCode.ERR_BadVisBaseExtension, baseExtensionLocations[baseExtension], this, baseExtension);
+                }
+
+                if (baseExtension.HasFileLocalTypes() && !this.HasFileLocalTypes())
+                {
+                    diagnostics.Add(ErrorCode.ERR_FileTypeBase, baseExtensionLocations[baseExtension], baseExtension, this);
+                }
+            }
+
+            baseExtensionLocations.Free();
+
+            diagnostics.Add(Locations.FirstOrNone(), useSiteInfo);
+
+            Debug.Assert(anyExplicit is not null);
+            return new ExtensionInfo(underlyingType, baseExtensions, isExplicit: anyExplicit.Value);
+        }
+
+        /// <summary> Bind the base extensions for one part of a partial extension.</summary>
+        private ExtensionInfo MakeOneDeclaredExtensionInfo(ConsList<TypeSymbol> basesBeingResolved, SingleTypeDeclaration decl, BindingDiagnosticBag diagnostics, out bool sawUnderlyingType)
+        {
+            var syntax = (ExtensionDeclarationSyntax)decl.SyntaxReference.GetSyntax();
+            TypeSymbol? partUnderlyingType = null;
+            if (GetUnderlyingTypeSyntax(decl) is { } underlyingTypeSyntax)
+            {
+                sawUnderlyingType = true;
+                var location = underlyingTypeSyntax.Location;
+                var underlyingTypeBinder = this.DeclaringCompilation.GetBinder(underlyingTypeSyntax);
+                underlyingTypeBinder = adjustBinder(syntax, underlyingTypeBinder);
+                var underlyingTypeWithAnnotations = underlyingTypeBinder.BindType(underlyingTypeSyntax, diagnostics, basesBeingResolved);
+
+                TypeSymbol underlyingType = underlyingTypeWithAnnotations.Type;
+                checkStatic(diagnostics, location, underlyingType);
+                if (IsRestrictedExtensionUnderlyingType(underlyingTypeWithAnnotations))
```

Comment:
>IsRestrictedExtensionUnderlyingType(underlyingTypeWithAnnotations) [](http://example.com/codeflow?start=20&length=66)

I think we should not drop type when it is null annotated. Report an error - yes, drop - no.